src/golang/lib/airflow/sync.go

```diff
@@ -0,0 +1,277 @@
+	"github.com/aqueducthq/aqueduct/lib/vault"
+	"github.com/aqueducthq/aqueduct/lib/workflow/operator/connector/auth"
+	"github.com/aqueducthq/aqueduct/lib/workflow/utils"
+	"github.com/dropbox/godropbox/errors"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+// SyncWorkflowDags syncs all dags in `workflowDagIds` with any new
+// Airflow dag runs since the last sync. It returns an error, if any.
+func SyncWorkflowDags(
+	ctx context.Context,
+	workflowDagIds []uuid.UUID,
+	workflowReader workflow.Reader,
+	workflowDagReader workflow_dag.Reader,
+	operatorReader operator.Reader,
+	artifactReader artifact.Reader,
+	workflowDagEdgeReader workflow_dag_edge.Reader,
+	workflowDagResultReader workflow_dag_result.Reader,
+	workflowDagResultWriter workflow_dag_result.Writer,
+	operatorResultWriter operator_result.Writer,
+	artifactResultWriter artifact_result.Writer,
+	notificationWriter notification.Writer,
+	userReader user.Reader,
+	vault vault.Vault,
+	db database.Database,
+) error {
+	// Read each workflow dag from the database that needs to be synced
+	dbDags := make([]workflow_dag.DBWorkflowDag, 0, len(workflowDagIds))
+	for _, workflowDagId := range workflowDagIds {
+		dbDag, err := utils.ReadWorkflowDagFromDatabase(
+			ctx,
+			workflowDagId,
+			workflowReader,
+			workflowDagReader,
+			operatorReader,
+			artifactReader,
+			workflowDagEdgeReader,
+			db,
+		)
+		if err != nil {
+			return err
+		}
+
+		dbDags = append(dbDags, *dbDag)
+	}
+
+	for _, dbDag := range dbDags {
+		if err := syncWorkflowDag(
+			ctx,
+			&dbDag,
+			workflowReader,
+			workflowDagReader,
+			workflowDagResultReader,
+			workflowDagResultWriter,
+			operatorResultWriter,
+			artifactResultWriter,
+			notificationWriter,
+			userReader,
+			vault,
+			db,
+		); err != nil {
+			log.Errorf("Unable to sync with Airflow for WorkflowDag %v: %v", dbDag.Id, err)
+		}
+	}
+
+	return nil
+}
+
+// syncWorkflowDag fetches the latest runs from Airflow for the workflow dag
+// specified and populates the database with the results.
+// It returns an error, if any.
+func syncWorkflowDag(
+	ctx context.Context,
+	dbDag *workflow_dag.DBWorkflowDag,
+	workflowReader workflow.Reader,
+	workflowDagReader workflow_dag.Reader,
+	workflowDagResultReader workflow_dag_result.Reader,
+	workflowDagResultWriter workflow_dag_result.Writer,
+	operatorResultWriter operator_result.Writer,
+	artifactResultWriter artifact_result.Writer,
+	notificationWriter notification.Writer,
+	userReader user.Reader,
+	vault vault.Vault,
+	db database.Database,
+) error {
+	// Read Airflow credentials from vault
+	authConf, err := auth.ReadConfigFromSecret(
+		ctx,
+		dbDag.EngineConfig.AirflowConfig.IntegrationId,
+		vault,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create Airflow API client
+	cli, err := newClient(ctx, authConf)
+	if err != nil {
+		return err
+	}
+
+	// Get all Airflow DAG runs for `dag`
+	// TODO: Get around Airflow response limit
```

Comment:
Create Linear TASK

Reply:
https://linear.app/aqueducthq/issue/ENG-1531/improve-airflow-syncing-to-deal-with-api-limit-of-dag-runs-returned-in